core/examples/repl.rs

```diff
@@ -0,0 +1,465 @@
+use deno_core::cdp;
+use deno_core::error::AnyError;
+use deno_core::futures::channel::mpsc::UnboundedReceiver as FUnboundedReceiver;
+use deno_core::serde_json;
+use deno_core::serde_json::Value;
+use deno_core::FsModuleLoader;
+use deno_core::JsRuntime;
+use deno_core::LocalInspectorSession;
+use deno_core::PollEventLoopOptions;
+use deno_core::RuntimeOptions;
+use deno_unsync::spawn_blocking;
+use futures::FutureExt;
+use futures::StreamExt;
+use parking_lot::Mutex;
+use rustyline::error::ReadlineError;
+use rustyline::CompletionType;
+use rustyline::Config;
+use rustyline::Editor;
+use rustyline_derive::Completer;
+use rustyline_derive::Helper;
+use rustyline_derive::Highlighter;
+use rustyline_derive::Hinter;
+use rustyline_derive::Validator;
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering::Relaxed;
+use std::sync::Arc;
+use tokio::sync::mpsc::channel;
+use tokio::sync::mpsc::unbounded_channel;
+use tokio::sync::mpsc::Receiver;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio::sync::mpsc::UnboundedSender;
+
+fn main() -> Result<(), Error> {
+  let js_runtime = JsRuntime::new(RuntimeOptions {
+    module_loader: Some(Rc::new(FsModuleLoader)),
+    is_main: true,
+    ..Default::default()
+  });
+
+  let runtime = tokio::runtime::Builder::new_current_thread()
+    .enable_all()
+    .build()?;
+
+  let future = async move {
+    let mut repl_session = ReplSession::initialize(js_runtime).await?;
+    let mut rustyline_channel = rustyline_channel();
+    let helper = EditorHelper {
+      context_id: repl_session.context_id,
+      sync_sender: rustyline_channel.0,
+    };
+    let editor = ReplEditor::new(helper)?;
+
+    loop {
+      let line = read_line_and_poll_session(
+        &mut repl_session,
+        &mut rustyline_channel.1,
+        editor.clone(),
+      )
+      .await;
+      match line {
+        Ok(line) => {
+          editor.set_should_exit_on_interrupt(false);
+          let output = repl_session.evaluate_line_and_get_output(&line).await;
+
+          print!("{}", output);
+        }
+        Err(ReadlineError::Interrupted) => {
+          if editor.should_exit_on_interrupt() {
+            break;
+          }
+          editor.set_should_exit_on_interrupt(true);
+          println!("press ctrl+c again to exit");
+          continue;
+        }
+        Err(ReadlineError::Eof) => {
+          break;
+        }
+        Err(err) => {
+          println!("Error: {err:?}");
+          break;
+        }
+      };
+    }
+    Ok(())
+  };
+  runtime.block_on(future)
+}
+
+pub enum EvaluationOutput {
+  Value(String),
+  Error(String),
+}
+
+impl std::fmt::Display for EvaluationOutput {
+  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+    match self {
+      EvaluationOutput::Value(value) => f.write_str(value),
+      EvaluationOutput::Error(value) => f.write_str(value),
+    }
+  }
+}
```

Comment:
seems excessive for an example

Reply:
It's copy-pasted from `cli/tools/repl`. I'm trying to slim it down as much as possible so we can turn it into a runtime API to provide `node:repl` polyfill.

Reply:
Plus see #412 - Matt and I have been discussing this and we want to do a bin target to be able to test changes quickly without having to rebuilt `deno` itself (because iterating on `deno_core` is waaaaaay faster).